File: src/Starfield2026.BgEditor/backend/src/lib/ohana/Containers/GARC.ts
// Ported from Containers\GARC.cs
// GARC archive reader (Game Archive Container)

import { BinaryReader } from '../Core/BinaryReader.js';
import { OContainer, OFileEntry } from './OContainer.js';
import { FileIO } from '../Core/FileIO.js';

export class GARC {
  /**
   * Reads a GARC archive from a file path.
   */
  static loadFile(fileName: string): OContainer {
    return GARC.load(BinaryReader.fromFile(fileName));
  }

  /**
   * Reads a GARC archive from a BinaryReader.
   */
  static load(input: BinaryReader): OContainer {
    const output = new OContainer();
    output.data = input;

    const garcMagic = input.readMagic(4);
    const garcLength = input.readUInt32();
    const endian = input.readUInt16();
    const version = input.readUInt16(); // 0x400
    const sectionCount = input.readUInt32();
    const dataOffset = input.readUInt32();
    const decompressedLength = input.readUInt32();
    const compressedLength = input.readUInt32();

    input.seek(garcLength); // This is just the header "GARC" blk len, not the entire file

    // File Allocation Table Offsets
    const fatoPosition = input.position;
    const fatoMagic = input.readMagic(4);
    const fatoLength = input.readUInt32();
    const fatoEntries = input.readUInt16();
    input.readUInt16(); // 0xffff = Padding?

    const fatbPosition = fatoPosition + fatoLength;

    for (let i = 0; i < fatoEntries; i++) {
      input.seek(fatoPosition + 0xc + i * 4);
      input.seek(input.readUInt32() + fatbPosition + 0xc);

      const flags = input.readUInt32();

      let folder = '';
      if (flags !== 1) folder = `folder_${String(i).padStart(5, '0')}/`;

      for (let bit = 0; bit < 32; bit++) {
        if ((flags & (1 << bit)) > 0) {
          const startOffset = input.readUInt32();
          const endOffset = input.readUInt32();
          const length = input.readUInt32();

          const position = input.position;

          input.seek(startOffset + dataOffset);

          // Only read the first few bytes needed for magic detection (not the full entry)
          const peekLen = Math.min(length, 16);
          const peek = input.readBytes(peekLen);

          const isCompressed = peek.length > 0 ? peek[0] === 0x11 : false;
          const extension = FileIO.getExtension(peek, isCompressed ? 5 : 0);
          const name = folder + `file_${String(flags === 1 ? i : bit).padStart(5, '0')}${extension}`;

          // Add the file to the container list
          const entry: OFileEntry = {
            name,
            data: null,
            loadFromDisk: true,
            fileOffset: startOffset + dataOffset,
            fileLength: length,
            doDecompression: isCompressed,
          };
          output.content.push(entry);

          input.seek(position);
        }
      }
    }

    return output;
  }
}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Spica/Cli/Formats/GARC.cs
using System.IO;
using System.Text;

namespace SpicaCli.Formats
{
    /// <summary>
    /// GARC (Game ARChive) container reader.
    /// Magic: "CRAG" (GARC reversed). Used in Pokemon 3DS ROMs.
    /// </summary>
    public static class GARC
    {
        public struct GARCEntry
        {
            public long Offset;
            public int Length;
        }

        /// <summary>
        /// Check if stream starts with GARC magic "CRAG".
        /// </summary>
        public static bool IsGARC(Stream stream)
        {
            if (stream.Length < 4) return false;

            long pos = stream.Position;
            byte[] magic = new byte[4];
            stream.Read(magic, 0, 4);
            stream.Seek(pos, SeekOrigin.Begin);

            return Encoding.ASCII.GetString(magic) == "CRAG";
        }

        /// <summary>
        /// Read all entry offsets and lengths from a GARC archive.
        /// </summary>
        public static GARCEntry[] GetEntries(Stream data)
        {
            BinaryReader input = new BinaryReader(data);

            // GARC header
            string magic = Encoding.ASCII.GetString(input.ReadBytes(4)); // "CRAG"
            uint garcHeaderLength = input.ReadUInt32();
            ushort endian = input.ReadUInt16();
            ushort version = input.ReadUInt16();
            uint sectionCount = input.ReadUInt32();
            uint dataOffset = input.ReadUInt32();
            uint decompressedLength = input.ReadUInt32();
            uint compressedLength = input.ReadUInt32();

            // Seek past GARC header to FATO section
            data.Seek(garcHeaderLength, SeekOrigin.Begin);

            // FATO (File Allocation Table Offsets)
            long fatoPosition = data.Position;
            string fatoMagic = Encoding.ASCII.GetString(input.ReadBytes(4)); // "OTAF"
            uint fatoLength = input.ReadUInt32();
            ushort fatoEntries = input.ReadUInt16();
            input.ReadUInt16(); // padding 0xFFFF

            long fatbPosition = fatoPosition + fatoLength;

            var entries = new List<GARCEntry>();

            for (int i = 0; i < fatoEntries; i++)
            {
                data.Seek(fatoPosition + 0xC + i * 4, SeekOrigin.Begin);
                uint fatbOffset = input.ReadUInt32();

                data.Seek(fatbPosition + 0xC + fatbOffset, SeekOrigin.Begin);

                uint flags = input.ReadUInt32();

                for (int bit = 0; bit < 32; bit++)
                {
                    if ((flags & (1 << bit)) > 0)
                    {
                        uint startOffset = input.ReadUInt32();
                        uint endOffset = input.ReadUInt32();
                        uint length = input.ReadUInt32();

                        entries.Add(new GARCEntry
                        {
                            Offset = startOffset + dataOffset,
                            Length = (int)length
                        });
                    }
                }
            }

            return entries.ToArray();
        }

        /// <summary>
        /// Read a single entry's data from the GARC archive.
        /// </summary>
        public static byte[] ReadEntry(Stream data, GARCEntry entry)
        {
            data.Seek(entry.Offset, SeekOrigin.Begin);
            byte[] buffer = new byte[entry.Length];
            data.Read(buffer, 0, buffer.Length);
            return buffer;
        }
    }
}
